runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -7176,29 +7176,39 @@ J9::Z::TreeEvaluator::VMifInstanceOfEvaluator(TR::Node * node, TR::CodeGenerator
    }
 
 /**
- * Generates a quick runtime test for valueType node and in case if node is of valueType, generates a branch to helper call
+ * Generates a quick runtime test for valueType node and in case if node is of valueType or valueBased, generates a branch to helper call
  *
  * @param node monent/exit node
+ * @param tempReg temporary register used for loading the constant flag
+ * @param deps register dependency
  * @param mergeLabel Label pointing to merge point
  * @param helperCallLabel Label pointing to helper call dispatch sequence.
  * @param cg Codegenerator object
  * @return Returns a register containing objectClassPointer
  */
 static TR::Register*
-generateCheckForValueTypeMonitorEnterOrExit(TR::Node *node, TR::LabelSymbol* mergeLabel, TR::LabelSymbol *helperCallLabel, TR::CodeGenerator *cg)
+generateCheckForValueMonitorEnterOrExit(TR::Node *node, TR::Register *tempReg, TR::RegisterDependencyConditions *deps, TR::LabelSymbol* mergeLabel,
+                                        TR::LabelSymbol *helperCallLabel, TR::CodeGenerator *cg)
    {
    TR::Register *objReg = cg->evaluate(node->getFirstChild());
    TR::Register *objectClassReg = cg->allocateRegister();
 
    TR::TreeEvaluator::genLoadForObjectHeadersMasked(cg, node, objectClassReg, generateS390MemoryReference(objReg, TR::Compiler->om.offsetOfObjectVftField(), cg), NULL);
 
-   // Currently the value of J9ClassIsValueType flag maps to the second (low order) byte of the class flags field. To
-   // be able to use the TM instruction we need to ensure we safe guard against changes of this value since we have to
-   // hardcode the offset from the class flags field to reach the byte corresponding to the bit we want to test.
-   static_assert(static_cast<uint32_t>(J9ClassIsValueType & 0xFF00) == static_cast<uint32_t>(J9ClassIsValueType), "Expecting J9ClassIsValueType to be within second byte of classFlags");
+   TR::MemoryReference *classFlagsMemRef = generateS390MemoryReference(objectClassReg, static_cast<uint32_t>(static_cast<TR_J9VMBase *>(cg->comp()->fe())->getOffsetOfClassFlags()), cg);
 
-   TR::MemoryReference *classFlagsMemRef = generateS390MemoryReference(objectClassReg, static_cast<uint32_t>(static_cast<TR_J9VMBase *>(cg->comp()->fe())->getOffsetOfClassFlags()) + 2, cg);
-   generateSIInstruction(cg, TR::InstOpCode::TM, node, classFlagsMemRef, J9ClassIsValueType >> 8);
+   // J9_CLASS_DISALLOWS_LOCKING_FLAGS is currently 0x40400 (J9ClassIsValueType | J9ClassIsValueBased)
+   if(cg->comp()->target().is64Bit())
+      {
+      // TODO: should deps be passed to genLoadLongConstant 
+      genLoadLongConstant(cg, node, J9_CLASS_DISALLOWS_LOCKING_FLAGS, tempReg, NULL, deps, NULL);
+      generateRXInstruction(cg, TR::InstOpCode::NG, node, tempReg, classFlagsMemRef);
+      }
+   else
+      {
+      generateLoad32BitConstant(cg, node, J9_CLASS_DISALLOWS_LOCKING_FLAGS, tempReg, true, NULL, deps, NULL);
```

Comment:
I would change this to 
`generateLoad32BitConstant(cg, node, J9_CLASS_DISALLOWS_LOCKING_FLAGS, tempReg, false);` as you do not need to set any condition code[1].


[1]. https://github.com/eclipse/omr/blob/2ccca9b155fa854e0b2606f8c8aa3c20c1645801/compiler/z/codegen/OMRTreeEvaluator.cpp#L127-L182

Reply:
>When you are anding the classFlags field with the constant value, I think you are using wrong set of instructions for 64 Bit. It should be same as 32 bit as classFlags within J9Class is 32-Bit field [1]. 

I think I see what you mean here. I followed the example of how `J9AccClassRAMArray` (0x10000) is checked in Z [1]. But `classDepthAndFlags` is `UDATA` 64 bit on Z and `classFlags` is `U_32` 32 bit. 

> So the current comparison would be incorrect as it will AND the J9_CLASS_DISALLOWS_LOCKING_FLAGS with next field in J9Class after classFlags. I would just remove the check for platform size here.

I'm trying to understand this part. `genLoadLongConstant() `inserts an instruction `LGFI`, does it mean J9_CLASS_DISALLOWS_LOCKING_FLAGS is loaded to the higher 32 bit after `LGFI`? That's why `TR::InstOpCode::NG` will compare J9_CLASS_DISALLOWS_LOCKING_FLAGS to the higher 32 bit off from the memory referenced by`classFlagsMemRef`, which is the next field of J9Class? 

[1]
https://github.com/eclipse/openj9/blob/70bfd3c9c373d2f13ba7e95a586a775341f9ed48/runtime/compiler/z/codegen/J9TreeEvaluator.cpp#L9296-L9304

Reply:
>I'm trying to understand this part. genLoadLongConstant() inserts an instruction LGFI, does it mean J9_CLASS_DISALLOWS_LOCKING_FLAGS is loaded to the higher 32 bit after LGFI? 

No, Let's see the example where `J9_CLASS_DISALLOWS_LOCKING_FLAGS = 0x40400`

LGFI R1, J9_CLASS_DISALLOWS_LOCKING_FLAGS // R1=0x00000000 00040400
NG   R1, @(J9Class + offsetOf(classFlags)) 

Keep in mind that S390 is big endian platform so the memory contents pointed by the memory reference in the above  `NG` instruction will have classFlags in the top half while bottom half will have other data. So the actual class Flags in the J9Class will be anded with upper Half of the R1 which is 0x00000000 and constant `J9_CLASS_DISALLOWS_LOCKING_FLAGS` will be anded with the next field of J9Class * after `classFlags`.

Reply:
Got it, thanks very much for the explanation @r30shah!